### PR TITLE
Add mysql secure connection function (closes #277)

### LINKED ISSUE #277: [Feature Request]mysql secure connection (TLS)
- [x] I certify that I have read the contributing guidelines and I acknowledge if I don't follow the format below or I don't check this box, my issue will be closed immediately without any notice.

**Is your feature request related to a problem? Please describe.**
trojan and the database of trojan panel perhaps exist on separate networks and they are in a high-risk network they could be sniff by gfw , I assume tls protocol could sloves this risk.
**Is this problem relevant to what trojan should care about?**
Trojan is a protocol implementation, not a full-fledged proxy client. Features such as custom routing will not be accepted.
Just about secure.
**Describe the solution you'd like**
the libmysqlclient has a mysql_ssl_set function.
And add a config about where is certificate authoritie if use trust third-party ca user can inputs /etc/ssl/certs/ca-certificates.crt(in debian).
**Describe alternatives you've considered**
maybe use mysql odbc?
**Additional context**

### PR DESCRIPTION
**Pull request by mem0rz**
add mysql secure connection function

### CODE REVIEW

File: src/core/authenticator.cpp [hunk 1]
```diff
@@ -27,6 +27,13 @@ using namespace std;
 Authenticator::Authenticator(const Config &config) {
     mysql_init(&con);
     Log::log_with_date_time("connecting to MySQL server " + config.mysql.server_addr + ':' + to_string(config.mysql.server_port), Log::INFO);
+    if (config.mysql.cafile.c_str() == NULL) {
+	} else if (config.mysql.tls_version.c_str() == NULL) {
+		mysql_ssl_set(&con, NULL, NULL, config.mysql.cafile.c_str(), NULL, NULL);
+	} else  {
+		mysql_ssl_set(&con, NULL, NULL, config.mysql.cafile.c_str(), NULL, NULL);
+		mysql_optionsv(&con, MARIADB_OPT_TLS_VERSION, config.mysql.tls_version.c_str());
+    }
```
Comment (wongsyrone) [inline src/core/authenticator.cpp]:
```suggestion
    if (config.mysql.cafile != "") {
        mysql_ssl_set(&con, NULL, NULL, config.mysql.cafile.c_str(), NULL, NULL);
    }
    if (config.mysql.tls_version == "") {
        mysql_optionsv(&con, MARIADB_OPT_TLS_VERSION, config.mysql.tls_version.c_str());
    }
```

File: src/core/config.cpp [hunk 2]
```diff
@@ -98,6 +98,8 @@ void Config::populate(const ptree &tree) {
     mysql.database = tree.get("mysql.database", string("trojan"));
     mysql.username = tree.get("mysql.username", string("trojan"));
     mysql.password = tree.get("mysql.password", string());
+    mysql.cafile = tree.get("mysql.cafile", string());
+	mysql.tls_version = tree.get("mysql.tls_version", string());
```
Comment (wongsyrone) [inline src/core/config.cpp]:
```suggestion
    mysql.cafile = tree.get("mysql.cafile", string());
    mysql.tls_version = tree.get("mysql.tls_version", string());
```

File: docs/authenticator.md [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # Authenticator
 
-Trojan servers can authenticate users according to not only passwords in the config file but also entries in a MySQL (MariaDB) database. To turn this functionality on, set `enabled` field in the MySQL config to `true` and correctly configure the server address and credentials, etc:
+Trojan servers can authenticate users according to not only passwords in the config file but also entries in a MySQL (MariaDB) database. To turn this functionality on, set `enabled` field in the MySQL config to `true` and correctly configure the server address and credentials,if you enable ssl connect to database you need fill the cafile(sure you can use 3rd-party trusted ca file in debian just like this "cafile":"/etc/ssl/certs/ca-certificates.crt", https://mariadb.com/kb/en/certificate-creation-with-openssl/#creating-a-certificate-authority-private-key-and-certificate) and tls_version(if have many tls version you need seprate them with comma like this "tls_version":"TLSv1.2,TLSv1.3"), etc:
```
Comment (wongsyrone) [inline docs/authenticator.md]: We should add a reference to https://mariadb.com/kb/en/mysql_optionsv/

**Comment by GreaterFire** [comment]
Thanks! I'll review soon.

File: src/core/authenticator.cpp [hunk 4]
```diff
@@ -27,6 +27,12 @@ using namespace std;
 Authenticator::Authenticator(const Config &config) {
     mysql_init(&con);
     Log::log_with_date_time("connecting to MySQL server " + config.mysql.server_addr + ':' + to_string(config.mysql.server_port), Log::INFO);
+    if (config.mysql.cafile != "") {
+        mysql_ssl_set(&con, NULL, NULL, config.mysql.cafile.c_str(), NULL, NULL);
+    }
+    if (config.mysql.tls_version == "") {
```
Comment (Chigusa0w0) [inline src/core/authenticator.cpp]: Do you really mean `tls_version == ""`?

Comment (cybmp3) [reply src/core/authenticator.cpp]: https://mariadb.com/kb/en/configuring-mariadb-connectorc-with-option-files/#tls_version

Comment (mem0rz) [reply src/core/authenticator.cpp]: My fault. I have fixed it.

File: src/core/authenticator.cpp [hunk 5]
```diff
@@ -27,6 +27,12 @@ using namespace std;
 Authenticator::Authenticator(const Config &config) {
     mysql_init(&con);
     Log::log_with_date_time("connecting to MySQL server " + config.mysql.server_addr + ':' + to_string(config.mysql.server_port), Log::INFO);
+    if (config.mysql.cafile != "") {
+        mysql_ssl_set(&con, NULL, NULL, config.mysql.cafile.c_str(), NULL, NULL);
```
Comment (GreaterFire) [inline src/core/authenticator.cpp:31]: According to https://dev.mysql.com/doc/refman/8.0/en/mysql-ssl-set.html,
```suggestion
        mysql_options(&con, MYSQL_OPT_SSL_CA, config.mysql.cafile.c_str());
```
seems simpler.

File: src/core/authenticator.cpp [hunk 6]
```diff
@@ -27,6 +27,12 @@ using namespace std;
...
+    }
+    if (config.mysql.tls_version != "") {
+        mysql_optionsv(&con, MARIADB_OPT_TLS_VERSION, config.mysql.tls_version.c_str());
```
Comment (GreaterFire) [inline src/core/authenticator.cpp]: `mysql_optionsv` and `MARIADB_OPT_TLS_VERSION` doesn't exist in MySQL.
```suggestion
        mysql_options(&con, MYSQL_OPT_TLS_VERSION, config.mysql.tls_version.c_str());
```

File: docs/authenticator.md (hunk 3 above)
Comment (GreaterFire) [inline docs/authenticator.md]:
```suggestion
Trojan servers can authenticate users according to not only passwords in the config file but also entries in a MySQL (MariaDB) database. To turn this functionality on, set `enabled` field in the MySQL config to `true` and correctly configure the server address, credentials, and etc. If you would like to connect to the database securely, you can to fill the `cafile` and/or the `tls_version` field (refer to [MySQL Documentation](https://dev.mysql.com/doc/refman/8.0/en/encrypted-connection-protocols-ciphers.html)):
```

**Review by GreaterFire — CHANGES_REQUESTED** [review]
IMO, we don't need the `tls_version` field as it can be configured elsewhere (https://dev.mysql.com/doc/refman/8.0/en/encrypted-connection-protocols-ciphers.html).

Generally LGTM. Just a bit of comments.

Could you test these changes?

**Pull request merged**